source/common/network/address_impl.h

```diff
@@ -85,21 +86,44 @@ class InstanceFactory {
 public:
   template <typename InstanceType, typename... Args>
   static StatusOr<InstanceConstSharedPtr> createInstancePtr(Args&&... args) {
+    const static bool generate_address_string = !Runtime::runtimeFeatureEnabled(
+        "envoy.restart_features.address_factory_lazy_create_address_string");
     absl::Status status;
     // Use new instead of make_shared here because the instance constructors are private and must be
     // called directly here.
-    std::shared_ptr<InstanceType> instance(new InstanceType(status, std::forward<Args>(args)...));
+    std::shared_ptr<InstanceType> instance(
+        new InstanceType(status, generate_address_string, std::forward<Args>(args)...));
     if (!status.ok()) {
       return status;
     }
     return instance;
   }
 };
 
+class IpInstance : public InstanceBase {
+public:
+  // InstanceBase
+  const std::string& asString() const override {
+    if (address_port_string_.empty()) {
+      const_cast<IpInstance*>(this)->populateAddressPortString();
```

Comment:
nit: if we make `address_port_string_` mutable, then we can avoid the cast.

Reply:
Then we need to make populateAddressPortString() const too, which seems counter intuitive with its name.

Reply:
How bout this. It looks like this class is using a protected member, which the style guide does not like. What if we make the member private and mutable. Then we can change `populateAddressPortString()` to `populateAddressPortString(std::string& str) const`. This way it will just populate the string provided and can be logically const. (And we avoid protected members). 

WDYT?

Reply:
The caller of populateAddressPortString() (in subclasses) still need to pass in the string member field so the sting still need to be protected though.

Reply:
Oh, clever! SGTM